### PR TITLE
Cryogenic anomalies & armor

### PR DESCRIPTION
**Pull request by JimKil3**
## What Does This PR Do
Adds cryogenic anomalies! These shoot out watcher beams in random directions and produce cooled gasses (CO2 and N2O), as well as freezing the floor around them and shooting out nanofrost randomly. If they're left for too long, they'll detonate in a flood of super-cold CO2. Deconstructing the core gives biotech 7. (Sprite courtesy of ATP Engineer)

Also adds the reactive gelidic armor, which blocks attacks and fires a cold beam back at the attacker.. and in random directions, sometimes. There's also a low chance that, on hit, the system will need to eject coolant - just like its anomaly counterpart.

Also also adds a Vetus Speculator mode for cryogenic anomalies (mostly done by Qwerty). Make sure to bring survival pens!

The special anomaly item will come in a future PR, but I don't want to touch balance too much in different areas with an item AND a reactive armor type.

## Why It's Good For The Game
More variety in anomalies & reactive armor is good, and this opens the way for cool (literally and figuratively) items in the future.

## Images of changes
image

## Testing
Spawned it, measured gasses, beat up skrells wearing the armor, etc.

## Changelog
:cl: JimKil3, Qwertytoforty, ATP Engineer
add: Added cryogenic anomalies and reactive gelidic armor
add: Added a cryogenic anomaly mode to the Vetus Speculator
tweak: Tweaked the pyroclastic anomaly alert to say "pyroclastic anomaly" rather than "atmospheric anomaly"
/:cl:

### CODE REVIEW

File: code/game/objects/effects/anomalies.dm [hunk 1]
```diff
@@ -353,6 +353,54 @@
...
+		shootAt(target)
+
+	if(prob(50))
+		var/turf/simulated/T = get_turf(src)
+		if(istype(T))
+			T.atmos_spawn_air(LINDA_SPAWN_COLD | LINDA_SPAWN_N2O | LINDA_SPAWN_CO2, 20)
+
+	if(prob(10))
+		var/obj/effect/nanofrost_container/A = new /obj/effect/nanofrost_container(get_turf(src))
+		for(var/a=0, a<5, a++)
```
Comment (SteelSlayer) [inline code/game/objects/effects/anomalies.dm:380]:
```suggestion
		for(var/i in 1 to 5)
```

File: code/modules/research/anomaly/anomaly.dm [hunk 2]
```diff
@@ -88,9 +96,11 @@
 			armour_path = /obj/item/clothing/suit/armor/reactive/teleport
 		if(istype(I, /obj/item/assembly/signaler/anomaly/pyro))
 			armour_path = /obj/item/clothing/suit/armor/reactive/fire
+		if(istype(I, /obj/item/assembly/signaler/anomaly/cryo))
+			armour_path = /obj/item/clothing/suit/armor/reactive/cryo
```
Comment (SteelSlayer) [inline code/modules/research/anomaly/anomaly.dm:100]: It looks like the `anomaly_armour_types` list was never implemented properly. Could you change around this proc a bit so we don't need these ifs? I've written out what I think it should be (needs testing):
```dm
/obj/item/reactive_armour_shell/attackby(obj/item/I, mob/user, params)
	var/static/list/anomaly_armour_types = list(
		/obj/item/assembly/signaler/anomaly/grav = /obj/item/clothing/suit/armor/reactive/repulse,
		/obj/item/assembly/signaler/anomaly/flux = /obj/item/clothing/suit/armor/reactive/tesla,
		/obj/item/assembly/signaler/anomaly/bluespace = /obj/item/clothing/suit/armor/reactive/teleport,
		/obj/item/assembly/signaler/anomaly/pyro = /obj/item/clothing/suit/armor/reactive/fire,
		/obj/item/assembly/signaler/anomaly/cryo = /obj/item/clothing/suit/armor/reactive/cryo,
		/obj/item/assembly/signaler/anomaly/vortex = /obj/item/clothing/suit/armor/reactive/stealth
		)

	if(istype(I, /obj/item/assembly/signaler/anomaly))
		var/obj/item/assembly/signaler/anomaly/A = I
		var/armour_path = anomaly_armour_types[A.type]
		if(!armour_path)
			armour_path = /obj/item/clothing/suit/armor/reactive/stealth //Fallback
		to_chat(user, "<span class='notice'>You insert [A] into the chest plate, and the armor gently hums to life.</span>")
		new armour_path(get_turf(src))
		qdel(src)
		qdel(A)
	return ..()
```
(Notice the added `/obj/item/assembly/signaler/anomaly/vortex = /obj/item/clothing/suit/armor/reactive/stealth` entry in the list.)

Comment (JimKil3) [reply code/modules/research/anomaly/anomaly.dm]: Tested this and it works, implemented in commit `e1c6354` (steel review commit 2)

File: code/game/objects/effects/anomalies.dm [hunk 3]
```diff
@@ -353,6 +353,57 @@
 
 /////////////////////
 
+/obj/effect/anomaly/cryo
+	name = "cryogenic anomaly"
+	icon_state = "cryoanomaly"
+	desc = "Hope you brought a jacket!"
```
Comment (Henri215) [inline code/game/objects/effects/anomalies.dm:359]:
```suggestion
/obj/effect/anomaly/cryo
	name = "cryogenic anomaly"
	desc = "Hope you brought a jacket!"
	icon_state = "cryoanomaly"
```

File: code/game/objects/effects/anomalies.dm [hunk 4]
```diff
@@ -353,6 +353,57 @@
...
+	aSignal = /obj/item/assembly/signaler/anomaly/cryo
+
+/obj/effect/anomaly/cryo/anomalyEffect()
+	..()
+
+	var/list/turf_targets = list()
+	for(var/turf/T in oview(get_turf(src), 7))
+		turf_targets += T
+
+	for(var/I in 1 to rand(1,3))
```
Comment (Henri215) [inline code/game/objects/effects/anomalies.dm:369]:
```suggestion
	for(var/I in 1 to rand(1, 3))
```

File: code/game/objects/effects/anomalies.dm [hunk 5]
```diff
@@ -353,6 +353,57 @@
...
+			sleep(2)
+		A.Smoke()
+
+/obj/effect/anomaly/cryo/proc/shootAt(atom/movable/target)
+	var/turf/T = get_turf(src)
+	var/turf/U = get_turf(target)
+	if(!T || !U)
+		return
+	var/obj/item/projectile/temp/basilisk/O = new /obj/item/projectile/temp/basilisk(T)
+	playsound(get_turf(src), 'sound/weapons/taser2.ogg', 75, 1)
```
Comment (Henri215) [inline code/game/objects/effects/anomalies.dm:394]:
```suggestion
	playsound(get_turf(src), 'sound/weapons/taser2.ogg', 75, TRUE)
```

File: code/modules/clothing/suits/armor_suits.dm [hunk 6]
```diff
@@ -517,6 +517,54 @@
...
+		return FALSE
+	if(reaction_check(hitby) && use_power())
+		owner.visible_message("<span class='danger'>[src] blocks [attack_text], sending out freezing bolts!</span>")
+
+		for(var/mob/M in oview(get_turf(src), 7))
+			shootAt(M)
+
+		if(prob(10)) //rarely vent gasses
+			owner.visible_message("<span class='warning'>[src] vents excess coolant!</span>")
+			playsound(loc, 'sound/effects/refill.ogg', 50, 1)
```
Comment (Henri215) [inline code/modules/clothing/suits/armor_suits.dm:545]:
```suggestion
			playsound(loc, 'sound/effects/refill.ogg', 50, TRUE)
```

File: code/modules/clothing/suits/armor_suits.dm [hunk 7]
```diff
@@ -517,6 +517,54 @@
...
+		return TRUE
+	return FALSE
+
+/obj/item/clothing/suit/armor/reactive/cryo/proc/shootAt(atom/movable/target)
+	var/turf/T = get_turf(src)
+	var/turf/U = get_turf(target)
+	if(!T || !U)
+		return
+	var/obj/item/projectile/temp/basilisk/O = new /obj/item/projectile/temp/basilisk(T)
+	playsound(get_turf(src), 'sound/weapons/taser2.ogg', 75, 1)
```
Comment (Henri215) [inline code/modules/clothing/suits/armor_suits.dm:562]:
```suggestion
	playsound(get_turf(src), 'sound/weapons/taser2.ogg', 75, TRUE)
```

File: code/datums/status_effects/debuffs.dm [hunk 8]
```diff
@@ -936,3 +936,28 @@
...
+	var/damage = 0.75
+	var/source_UID
+
+/datum/status_effect/cryo_beam/on_creation(mob/living/new_owner, mob/living/source)
+	. = ..()
+	source_UID = source.UID()
+
+/datum/status_effect/cryo_beam/tick()
+	var/mob/living/simple_animal/hostile/megafauna/ancient_robot/attacker = locateUID(source_UID)
+	if(owner in view(attacker, 8))
```
Comment (lewcc) [inline code/datums/status_effects/debuffs.dm:954]: Invert this condition and make the qdel(src) early return

File: code/game/objects/effects/anomalies.dm [hunk 9]
```diff
@@ -353,6 +353,57 @@
...
+	playsound(get_turf(src), 'sound/weapons/taser2.ogg', 75, TRUE)
+	O.current = T
+	O.yo = U.y - T.y
+	O.xo = U.x - T.x
+	O.fire()
+
+/obj/effect/anomaly/cryo/detonate()
+	var/turf/simulated/T = get_turf(src)
+	if(istype(T) && drops_core)
+		T.atmos_spawn_air(LINDA_SPAWN_COLD | LINDA_SPAWN_CO2, 1000)
```
Comment (lewcc) [inline code/game/objects/effects/anomalies.dm:403]: I'm a little concerned about spawning this much gas on lavaland

File: code/modules/mob/living/simple_animal/hostile/megafauna/ancient_robot.dm [hunk 10]
```diff
@@ -518,6 +527,14 @@ Difficulty: Hard
 				for(var/atom/A in T)
 					A.ex_act(3) //Body is immune to explosions of this strength.
 				T.ex_act(3)
+			if(mode == CRYO)
+				var/turf/simulated/S = get_turf(src)
+				S.MakeSlippery(TURF_WET_ICE)
+				for(var/turf/T in range (1,src))
```
Comment (lewcc) [inline code/modules/mob/living/simple_animal/hostile/megafauna/ancient_robot.dm:533]:
```suggestion
				for(var/turf/T in range (1, src))
```

File: code/game/objects/effects/anomalies.dm (hunk 9 above)
Comment (JimKil3) [reply code/game/objects/effects/anomalies.dm:403]: I brought this up with Qwerty when we were working on the vetus stuff and they said it'd probably be fine

Comment (Qwertytoforty) [reply code/game/objects/effects/anomalies.dm:403]: @lewcc The drops_core check means that vetus anomalies will not do the 1000 mole dentonation

Comment (JimKil3) [reply code/game/objects/effects/anomalies.dm:403]: ^^ That too

**Review by S34NW — APPROVED** [review]

**Pull request merged**